File: Fuentes/Configuracion Correo/Escritorio/Principal.vb
Imports System.IO

Public Class Principal
     
    Dim proveedores As New EntidadesConfiguracionCorreo.ProveedoresCorreo
    Dim configuracion As New EntidadesConfiguracionCorreo.ConfiguracionCorreo
    Public datosEmpresa As New LogicaConfiguracionCorreo.DatosEmpresa()
    Public datosUsuario As New LogicaConfiguracionCorreo.DatosUsuario()
    Public datosArea As New LogicaConfiguracionCorreo.DatosArea()
    Public pintado As Boolean = False
    Dim ejecutarProgramaPrincipal As New ProcessStartInfo()
    Public tieneDatos As Boolean = False
    Public estaCerrando As Boolean = False

    Public esDesarrollo As Boolean = False

#Region "Eventos"

    Private Sub Principal_FormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed

        Me.Cursor = Cursors.WaitCursor
        Dim nombrePrograma As String = "PrincipalBerry"
        AbrirPrograma(nombrePrograma, True)
        System.Threading.Thread.Sleep(5000)
        Me.Cursor = Cursors.Default

    End Sub

    Private Sub Principal_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing

        Me.estaCerrando = True
        Desvanecer()

    End Sub

    Private Sub Principal_Load(sender As Object, e As EventArgs) Handles MyBase.Load

        Centrar()
        AsignarTooltips()
        ConfigurarConexiones()

    End Sub

    Private Sub Principal_Shown(sender As Object, e As EventArgs) Handles Me.Shown

        Me.Cursor = Cursors.AppStarting
        Me.Enabled = False
        If (Not ValidarAccesoTotal()) Then
            Salir()
        End If
        CargarEncabezados()
        CargarComboProveedores()
        CargarConfiguracion()
        Me.Enabled = True
        Me.Cursor = Cursors.Default

    End Sub

    Private Sub btnSalir_Click(sender As Object, e As EventArgs) Handles btnSalir.Click

        Salir()

    End Sub

    Private Sub btnGuardar_Click(sender As Object, e As EventArgs) Handles btnGuardar.Click

        Guardar()

    End Sub

    Private Sub btnGuardar_MouseHover(sender As Object, e As EventArgs) Handles btnGuardar.MouseHover

        AsignarTooltips("Guardar.")

    End Sub

    Private Sub btnSalir_MouseHover(sender As Object, e As EventArgs) Handles btnSalir.MouseHover

        AsignarTooltips("Salir.")

    End Sub

    Private Sub Principal_Paint(sender As Object, e As PaintEventArgs) Handles MyBase.Paint

        If (Not Me.pintado) Then
            Me.pintado = True
            txtDireccion.Focus()
        End If

    End Sub

    Private Sub txtDireccion_KeyDown(sender As Object, e As KeyEventArgs) Handles txtDireccion.KeyDown

        If e.KeyCode = Keys.Enter Then
            If (Not String.IsNullOrEmpty(txtDireccion.Text)) Then
                txtContrasena.Focus()
            End If
        End If

    End Sub

    Private Sub txtContrasena_KeyDown(sender As Object, e As KeyEventArgs) Handles txtContrasena.KeyDown

        If e.KeyCode = Keys.Escape Then
            txtDireccion.Focus()
        ElseIf e.KeyCode = Keys.Enter Then
            If (Not String.IsNullOrEmpty(txtContrasena.Text)) Then
                txtAsunto.Focus()
            End If
        End If

    End Sub

    Private Sub txtAsunto_KeyDown(sender As Object, e As KeyEventArgs) Handles txtAsunto.KeyDown

        If e.KeyCode = Keys.Escape Then
            txtContrasena.Focus()
        ElseIf e.KeyCode = Keys.Enter Then
            If (Not String.IsNullOrEmpty(txtAsunto.Text)) Then
                txtMensaje.Focus()
            End If
        End If

    End Sub

    Private Sub txtMensaje_KeyDown(sender As Object, e As KeyEventArgs) Handles txtMensaje.KeyDown

        If e.KeyCode = Keys.Escape Then
            txtAsunto.Focus()
        ElseIf e.KeyCode = Keys.Enter Then
            If (Not String.IsNullOrEmpty(txtMensaje.Text)) Then
                btnGuardar.Focus()
            End If
        End If

    End Sub

    Private Sub pnlEncabezado_MouseHover(sender As Object, e As EventArgs) Handles pnlPie.MouseHover, pnlEncabezado.MouseHover, pnlCuerpo.MouseHover

        AsignarTooltips(String.Empty)

    End Sub

    Private Sub temporizador_Tick(sender As Object, e As EventArgs) Handles temporizador.Tick

        If (Me.estaCerrando) Then
            Desvanecer()
        End If

    End Sub

    Private Sub btnAyuda_Click(sender As Object, e As EventArgs) Handles btnAyuda.Click

        MostrarAyuda()

    End Sub

    Private Sub btnAyuda_MouseHover(sender As Object, e As EventArgs) Handles btnAyuda.MouseHover

        AsignarTooltips("Ayuda.")

    End Sub

#End Region

#Region "Métodos"

#Region "Genericos"

    Private Sub Salir()

        Application.Exit()

    End Sub

    Private Function ValidarAccesoTotal() As Boolean

        If ((Not datosUsuario.EAccesoTotal) Or (datosUsuario.EAccesoTotal = 0) Or (datosUsuario.EAccesoTotal = False)) Then
            MsgBox("No tienes permisos suficientes para acceder a este programa.", MsgBoxStyle.Information, "Permisos insuficientes.")
            Return False
        Else
            Return True
        End If

    End Function

    Private Sub MostrarAyuda()

        Dim pnlAyuda As New Panel()
        Dim txtAyuda As New TextBox()
        If (pnlContenido.Controls.Find("pnlAyuda", True).Count = 0) Then
            pnlAyuda.Name = "pnlAyuda" : Application.DoEvents()
            pnlAyuda.Visible = False : Application.DoEvents()
            pnlContenido.Controls.Add(pnlAyuda) : Application.DoEvents()
            txtAyuda.Name = "txtAyuda" : Application.DoEvents()
            pnlAyuda.Controls.Add(txtAyuda) : Application.DoEvents()
        Else
            pnlAyuda = pnlContenido.Controls.Find("pnlAyuda", False)(0) : Application.DoEvents()
            txtAyuda = pnlAyuda.Controls.Find("txtAyuda", False)(0) : Application.DoEvents()
        End If
        If (Not pnlAyuda.Visible) Then
            pnlCuerpo.Visible = False : Application.DoEvents()
            pnlAyuda.Visible = True : Application.DoEvents()
            pnlAyuda.Size = pnlCuerpo.Size : Application.DoEvents()
            pnlAyuda.Location = pnlCuerpo.Location : Application.DoEvents()
            pnlContenido.Controls.Add(pnlAyuda) : Application.DoEvents()
            txtAyuda.ScrollBars = ScrollBars.Both : Application.DoEvents()
            txtAyuda.Multiline = True : Application.DoEvents()
            txtAyuda.Width = pnlAyuda.Width - 10 : Application.DoEvents()
            txtAyuda.Height = pnlAyuda.Height - 10 : Application.DoEvents()
            txtAyuda.Location = New Point(5, 5) : Application.DoEvents()
            txtAyuda.Text = "Sección de Ayuda: " & vbNewLine & vbNewLine & "* Configuración de Correo: " & vbNewLine & "En esta parte se configura la dirección y contraseña de correo como parte de las credenciales de inicio de sesión del cual se enviarán las notificaciones a los usuarios. " & vbNewLine & "En asunto va pues el asunto del correo, y mensaje el mensaje del correo. En dado caso de que se configure el asunto y/o mensaje en blanco o vacios se les pondrá un asunto y/o mensaje por defecto al enviarlos." : Application.DoEvents()
            pnlAyuda.Controls.Add(txtAyuda) : Application.DoEvents()
        Else
            pnlCuerpo.Visible = True : Application.DoEvents()
            pnlAyuda.Visible = False : Application.DoEvents()
        End If

    End Sub

    Private Sub Desvanecer()

        temporizador.Interval = 10
        temporizador.Enabled = True
        temporizador.Start()
        If (Me.Opacity > 0) Then
            Me.Opacity -= 0.25 : Application.DoEvents()
        Else
            temporizador.Enabled = False
            temporizador.Stop()
        End If

    End Sub

    Private Sub Centrar()

        Me.CenterToScreen()
        'Me.Opacity = 0.97 'Está bien perro esto.  
        Me.Location = Screen.PrimaryScreen.WorkingArea.Location
        Me.Size = Screen.PrimaryScreen.WorkingArea.Size

    End Sub

    Private Sub AsignarTooltips()

        Dim tp As New ToolTip()
        tp.AutoPopDelay = 5000
        tp.InitialDelay = 0
        tp.ReshowDelay = 100
        tp.ShowAlways = True
        tp.SetToolTip(Me.btnSalir, "Salir.")
        tp.SetToolTip(Me.btnGuardar, "Guardar o Editar.")

    End Sub

    Private Sub AsignarTooltips(ByVal texto As String)

        lblDescripcionTooltip.Text = texto

    End Sub

    Public Sub ControlarSpreadEnter(ByVal spread As FarPoint.Win.Spread.FpSpread)

        Dim valor1 As FarPoint.Win.Spread.InputMap
        Dim valor2 As FarPoint.Win.Spread.InputMap
        valor1 = spread.GetInputMap(FarPoint.Win.Spread.InputMapMode.WhenAncestorOfFocused)
        valor1.Put(New FarPoint.Win.Spread.Keystroke(Keys.Enter, Keys.None), FarPoint.Win.Spread.SpreadActions.MoveToNextColumnWrap)
        valor1 = spread.GetInputMap(FarPoint.Win.Spread.InputMapMode.WhenFocused)
        valor1.Put(New FarPoint.Win.Spread.Keystroke(Keys.Enter, Keys.None), FarPoint.Win.Spread.SpreadActions.MoveToNextColumnWrap)
        valor2 = spread.GetInputMap(FarPoint.Win.Spread.InputMapMode.WhenFocused)
        valor2.Put(New FarPoint.Win.Spread.Keystroke(Keys.Escape, Keys.None), FarPoint.Win.Spread.SpreadActions.None)
        valor2 = spread.GetInputMap(FarPoint.Win.Spread.InputMapMode.WhenAncestorOfFocused)
        valor2.Put(New FarPoint.Win.Spread.Keystroke(Keys.Escape, Keys.None), FarPoint.Win.Spread.SpreadActions.None)

    End Sub

    Private Sub ConfigurarConexiones()

        If (Me.esDesarrollo) Then
            'baseDatos.CadenaConexionInformacion = "C:\\Berry-Agenda\\BD\\PODC\\Agenda.mdf"
            Me.datosUsuario.EId = 1
            Me.datosUsuario.EIdArea = 1
            Me.datosEmpresa.EId = 1
            Me.datosUsuario.EAccesoTotal = True
            'LogicaConfiguracionCorreo.DatosEmpresaPrincipal.instanciaSql = "ANDREW-MAC\SQLEXPRESS"
            LogicaConfiguracionCorreo.DatosEmpresaPrincipal.instanciaSql = "BERRY1-DELL\SQLEXPRESS2008"
            LogicaConfiguracionCorreo.DatosEmpresaPrincipal.usuarioSql = "AdminBerry"
            'LogicaConfiguracionCorreo.DatosEmpresaPrincipal.contrasenaSql = "@berry"
            LogicaConfiguracionCorreo.DatosEmpresaPrincipal.contrasenaSql = "@berry2017"
        Else
            'EntidadesActividades.BaseDatos.ECadenaConexionAgenda = datosEmpresa.EDirectorio & "\\Agenda.mdf"
            LogicaConfiguracionCorreo.DatosEmpresaPrincipal.ObtenerParametrosInformacionEmpresa()
            Me.datosEmpresa.ObtenerParametrosInformacionEmpresa()
            Me.datosUsuario.ObtenerParametrosInformacionUsuario()
            Me.datosArea.ObtenerParametrosInformacionArea()
        End If
        EntidadesConfiguracionCorreo.BaseDatos.ECadenaConexionAgenda = "Agenda"
        EntidadesConfiguracionCorreo.BaseDatos.ECadenaConexionCatalogo = "Catalogos"
        EntidadesConfiguracionCorreo.BaseDatos.ECadenaConexionInformacion = "Informacion"
        EntidadesConfiguracionCorreo.BaseDatos.AbrirConexionAgenda()
        EntidadesConfiguracionCorreo.BaseDatos.AbrirConexionCatalogo()
        EntidadesConfiguracionCorreo.BaseDatos.AbrirConexionInformacion()

    End Sub

    Private Sub CargarEncabezados()

        lblEncabezadoPrograma.Text = "Programa: " + Me.Text
        lblEncabezadoEmpresa.Text = "Empresa: " + datosEmpresa.ENombre
        lblEncabezadoUsuario.Text = "Usuario: " + datosUsuario.ENombre
        lblEncabezadoArea.Text = "Area: " + datosArea.ENombre

    End Sub

    Private Sub PonerFocoEnControl(ByVal c As Control)

        c.Focus()

    End Sub

    Private Sub AbrirPrograma(nombre As String, salir As Boolean)

        If (Me.esDesarrollo) Then
            Exit Sub
        End If
        ejecutarProgramaPrincipal.UseShellExecute = True
        ejecutarProgramaPrincipal.FileName = nombre & Convert.ToString(".exe")
        ejecutarProgramaPrincipal.WorkingDirectory = Application.StartupPath
        ejecutarProgramaPrincipal.Arguments = LogicaConfiguracionCorreo.DatosEmpresaPrincipal.idEmpresa.ToString().Trim().Replace(" ", "|") & " " & LogicaConfiguracionCorreo.DatosEmpresaPrincipal.activa.ToString().Trim().Replace(" ", "|") & " " & LogicaConfiguracionCorreo.DatosEmpresaPrincipal.instanciaSql.ToString().Trim().Replace(" ", "|") & " " & LogicaConfiguracionCorreo.DatosEmpresaPrincipal.rutaBd.ToString().Trim().Replace(" ", "|") & " " & LogicaConfiguracionCorreo.DatosEmpresaPrincipal.usuarioSql.ToString().Trim().Replace(" ", "|") & " " & LogicaConfiguracionCorreo.DatosEmpresaPrincipal.contrasenaSql.ToString().Trim().Replace(" ", "|") & " " & "Aquí terminan los de empresa principal, indice 7 ;)".Replace(" ", "|") & " " & datosEmpresa.EId.ToString().Trim().Replace(" ", "|") & " " & datosEmpresa.ENombre.Trim().Replace(" ", "|") & " " & datosEmpresa.EDescripcion.Trim().Replace(" ", "|") & " " & datosEmpresa.EDomicilio.Trim().Replace(" ", "|") & " " & datosEmpresa.ELocalidad.Trim().Replace(" ", "|") & " " & datosEmpresa.ERfc.Trim().Replace(" ", "|") & " " & datosEmpresa.EDirectorio.Trim().Replace(" ", "|") & " " & datosEmpresa.ELogo.Trim().Replace(" ", "|") & " " & datosEmpresa.EActiva.ToString().Trim().Replace(" ", "|") & " " & datosEmpresa.EEquipo.Trim().Replace(" ", "|") & " " & "Aquí terminan los de empresa, indice 18 ;)".Replace(" ", "|") & " " & datosUsuario.EId.ToString().Trim().Replace(" ", "|") & " " & datosUsuario.ENombre.Trim().Replace(" ", "|") & " " & datosUsuario.EContrasena.Trim().Replace(" ", "|") & " " & datosUsuario.ENivel.ToString().Trim().Replace(" ", "|") & " " & datosUsuario.EAccesoTotal.ToString().Trim().Replace(" ", "|") & " " & datosUsuario.EIdArea.ToString().Trim().Replace(" ", "|") & " " & "Aquí terminan los de usuario, indice 25 ;)".Replace(" ", "|") & " " & datosArea.EId.ToString().Trim().Replace(" ", "|") & " " & datosArea.ENombre.ToString().Trim().Replace(" ", "|") & " " & datosArea.EClave.ToString().Trim().Replace(" ", "|") & " " & "Aquí terminan los de area, indice 29 ;)".Replace(" ", "|")
        Try
            Process.Start(ejecutarProgramaPrincipal)
            If salir Then
                Application.Exit()
            End If
        Catch ex As Exception
            MessageBox.Show((Convert.ToString("No se puede abrir el programa principal en la ruta : " & ejecutarProgramaPrincipal.WorkingDirectory & "\") & nombre) & Environment.NewLine & Environment.NewLine & ex.Message, "Error.", MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try

    End Sub

#End Region

#Region "Todos"

    Private Sub CargarConfiguracion()

        Dim lista As New List(Of EntidadesConfiguracionCorreo.ConfiguracionCorreo)
        lista = configuracion.ObtenerListado()
        If lista.Count > 0 Then
            txtDireccion.Text = lista(0).EDireccion
            txtContrasena.Text = lista(0).EContrasena
            txtAsunto.Text = lista(0).EAsunto
            txtMensaje.Text = lista(0).EMensaje
            cbProveedores.SelectedValue = lista(0).EIdProveedor
            Me.tieneDatos = True
        Else
            Me.tieneDatos = False
        End If

    End Sub

    Private Sub Guardar()

        Dim direccion As String = txtDireccion.Text
        Dim contrasena As String = txtContrasena.Text
        Dim asunto As String = txtAsunto.Text
        Dim mensaje As String = txtMensaje.Text
        Dim idProveedor As Integer = cbProveedores.SelectedValue
        If ((Not String.IsNullOrEmpty(direccion)) And (Not String.IsNullOrEmpty(contrasena)) And (idProveedor > 0)) Then
            configuracion.EDireccion = direccion
            configuracion.EContrasena = contrasena
            configuracion.EAsunto = asunto
            configuracion.EMensaje = mensaje
            configuracion.EIdProveedor = idProveedor
            If (Me.tieneDatos) Then
                configuracion.Editar()
            Else
                configuracion.Guardar()
            End If
            MsgBox("Guardado finalizado.", MsgBoxStyle.ApplicationModal, "Finalizado.")
            CargarConfiguracion()
        End If

    End Sub

    Private Sub CargarComboProveedores()

        Dim lista As New List(Of EntidadesConfiguracionCorreo.ProveedoresCorreo)
        lista = proveedores.ObtenerListado()
        cbProveedores.DataSource = lista
        cbProveedores.ValueMember = "EId"
        cbProveedores.DisplayMember = "EDominio"

    End Sub

#End Region

#End Region
     
End Class

File: Fuentes/Activador/Escritorio/Principal.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Drawing.Drawing2D;
using System.Reflection;
using System.Data.Sql;

namespace Escritorio
{
    public partial class Principal : Form
    {
         
        Entidades.BaseDatos baseDatos = new Entidades.BaseDatos(); 
        Entidades.EmpresasPrincipal empresasPrincipal = new Entidades.EmpresasPrincipal();
        Entidades.Licencia licencia = new Entidades.Licencia(); 
        ProcessStartInfo ejecutarProgramaPrincipal = new ProcessStartInfo();
        public int numeroEmpresa; 
        public bool tieneParametros = false;
        public bool esInicioSesion = true;
        public int idEmpresaSesion = 0; public int idUsuarioSesion = 0; public int idModuloSesion = 1;
        public string nombrePrograma = string.Empty;
        public bool estaCerrando = false; public bool estaAbriendoPrograma = false;
        public int diasDePrueba = 15;
        public bool estaMostrado = false;

        public bool esDesarrollo = false;

        public Principal()
        {
            InitializeComponent();
        }
        
        #region Eventos

        private void Principal_Load(object sender, EventArgs e)
        {

            Centrar();
            AsignarTooltips();
            AsignarFocos();
            ConfigurarConexiones();
            CargarEncabezados();
            CargarTitulosEmpresa();

        }

        private void Principal_Shown(object sender, EventArgs e)
        {

            this.txtUsuario.Focus();
            this.estaMostrado = true;
            ValidarLicencia();

        }

        private void Principal_FormClosed(object sender, FormClosedEventArgs e)
        {

            Salir();

        }

        private void Principal_FormClosing(object sender, FormClosingEventArgs e)
        {

            this.estaCerrando = true;

        }
  
        private void txtUsuario_KeyDown(object sender, KeyEventArgs e)
        {

            if (e.KeyCode == Keys.Enter)
            {
                e.SuppressKeyPress = true;
                if (!string.IsNullOrEmpty(this.txtUsuario.Text))
                { 
                    this.txtContraseña.Focus();                
                }            
            } 

        }

        private void txtContraseña_KeyDown(object sender, KeyEventArgs e)
        { 

            if (e.KeyCode == Keys.Enter)
            {
                e.SuppressKeyPress = true;
                if (!string.IsNullOrEmpty(this.txtContraseña.Text))
                {
                    this.btnEntrar.Focus();
                }
            }
            else if (e.KeyCode == Keys.Escape)
            {
                this.txtUsuario.Focus();
            } 

        }

        private void btnIniciarSesion_Click(object sender, EventArgs e)
        { 

            ValidarSesion(); 

        }

        private void btnSalir_Click(object sender, EventArgs e)
        {

            SalirOVolver();

        }

        private void btnEntrar_MouseHover(object sender, EventArgs e)
        {

            AsignarTooltips("Entrar.");

        }

        private void btnSalir_MouseHover(object sender, EventArgs e)
        {

            if (this.esInicioSesion)
            {
                AsignarTooltips("Salir."); 
            }
            else
            {
                AsignarTooltips("Volver a Iniciar Sesión."); 
            }

        }
         
        private void pnlEncabezado_MouseHover(object sender, EventArgs e)
        {

            AsignarTooltips(string.Empty); 

        }

        private void pnlContenido_MouseHover(object sender, EventArgs e)
        {

            AsignarTooltips(string.Empty); 

        }

        private void pnlMenu_MouseHover(object sender, EventArgs e)
        {

            AsignarTooltips(string.Empty); 

        }

        private void pnlIniciarSesion_MouseHover(object sender, EventArgs e)
        {

            AsignarTooltips(string.Empty); 

        }

        private void pnlPie_MouseHover(object sender, EventArgs e)
        {

            AsignarTooltips(string.Empty); 

        }
         
        private void btnAyuda_MouseHover(object sender, EventArgs e)
        {

            AsignarTooltips("Ayuda.");

        }

        private void btnAyuda_Click(object sender, EventArgs e)
        {

            MostrarAyuda();

        }

        private void chkActivar_CheckedChanged(object sender, EventArgs e)
        {

            if (this.estaMostrado)
            {
                if (chkActivar.Checked)
                {
                    chkActivar.Text = "Activado";
                    Activar();
                }
                else
                {
                    chkActivar.Text = "Desactivado";
                    Desactivar();
                } 
            }

        }

        #endregion

        #region Métodos

        private void ValidarLicencia()
        {

            List<Entidades.Licencia> lista = new List<Entidades.Licencia>(); 
            lista = licencia.ObtenerListado();
            if (lista.Count > 0)
            {
                if (lista[0].EsPrueba)
                {
                    chkActivar.Checked = false;
                }
                else
                {
                    chkActivar.Checked = true;
                }
            }
            else
            {
                chkActivar.Checked = false;
            }

        }

        private void Salir() 
        {

            this.Cursor = Cursors.WaitCursor;
            if (this.estaAbriendoPrograma)
            {
                System.Threading.Thread.Sleep(10000);
            }
            Application.ExitThread();
            Application.Exit();
            this.Cursor = Cursors.Default;

        }

        private void SalirOVolver() 
        {

            if (this.esInicioSesion)
            {
                Application.Exit();
            }
            else
            {
                pnlContenido.BackgroundImage = global::Activador.Properties.Resources.Logo3;
                pnlContenido.BackgroundImageLayout = ImageLayout.Zoom;
                pnlContenido.BackColor = Color.DarkSlateGray;
                //pnlContenido.BackgroundImage = global::Principal.Properties.Resources.hiedra;
                //pnlContenido.BackgroundImageLayout = ImageLayout.Stretch;
                pnlPanelControl.Visible = false;
                pnlIniciarSesion.Visible = true;
                txtContraseña.Text = string.Empty;
                txtUsuario.Text = string.Empty;
                txtUsuario.Focus();
                this.esInicioSesion = true;
            }

        }

        private void MostrarAyuda()
        {

            Panel pnlAyuda = new Panel();
            TextBox txtAyuda = new TextBox();
            if (pnlContenido.Controls.Find("pnlAyuda", true).Count() == 0)
            {
                pnlAyuda.Name = "pnlAyuda"; Application.DoEvents();
                pnlAyuda.Visible = false; Application.DoEvents();
                pnlContenido.Controls.Add(pnlAyuda); Application.DoEvents();
                txtAyuda.Name = "txtAyuda"; Application.DoEvents();
                pnlAyuda.Controls.Add(txtAyuda); Application.DoEvents();
            }
            else
            {
                pnlAyuda = pnlContenido.Controls.Find("pnlAyuda", false).FirstOrDefault() as Panel; Application.DoEvents();
                txtAyuda = pnlAyuda.Controls.Find("txtAyuda", false).FirstOrDefault() as TextBox; Application.DoEvents();
            }
            if ((!pnlAyuda.Visible))
            {
                if (this.esInicioSesion)
                {
                    pnlIniciarSesion.Visible = false; Application.DoEvents();
                }
                else
                { 
                    pnlPanelControl.Visible = false; Application.DoEvents(); 
                }
                pnlAyuda.Visible = true; Application.DoEvents();
                pnlAyuda.Size = pnlPanelControl.Size; Application.DoEvents();
                pnlAyuda.Location = pnlPanelControl.Location; Application.DoEvents();
                pnlContenido.Controls.Add(pnlAyuda); Application.DoEvents();
                txtAyuda.ScrollBars = ScrollBars.Both; Application.DoEvents();
                txtAyuda.Multiline = true; Application.DoEvents();
                txtAyuda.Width = pnlAyuda.Width - 10; Application.DoEvents();
                txtAyuda.Height = pnlAyuda.Height - 10; Application.DoEvents();
                txtAyuda.Location = new Point(5, 5); Application.DoEvents();
                txtAyuda.Text = "Sección de Ayuda: " + System.Environment.NewLine + System.Environment.NewLine + "* Activador: " + System.Environment.NewLine + "Únicamente usuarios autorizados por Berry pueden utilizar este programa para activar el sistema. " + System.Environment.NewLine + "Saludos."; Application.DoEvents();
                pnlAyuda.Controls.Add(txtAyuda); Application.DoEvents();
            }
            else
            {
                if (this.esInicioSesion)
                {
                    pnlIniciarSesion.Visible = true; Application.DoEvents();
                }
                else 
                { 
                    pnlPanelControl.Visible = true; Application.DoEvents(); 
                }
                pnlAyuda.Visible = false; Application.DoEvents();
            }

        } 

        private void ValidarSesion()
        { 
             
            if (!string.IsNullOrEmpty(this.txtUsuario.Text) && !string.IsNullOrEmpty(this.txtContraseña.Text))
            {
                if ((txtUsuario.Text.ToUpper().Equals("Admin".ToUpper())) && (txtContraseña.Text.Equals("@berry2017")))
                { 
                    PermitirAcceso();
                }
                else
                {
                    MessageBox.Show("Credenciales no válidas.", "Datos incorrectos.", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                    txtUsuario.Focus();
                }
            }
            else
            {
                MessageBox.Show("Faltan datos.", "Faltan datos.", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                txtUsuario.Focus(); 
            }

        }

        private void PermitirAcceso()
        {

            pnlIniciarSesion.Visible = false; Application.DoEvents();
            pnlPanelControl.Visible = true; Application.DoEvents();
            this.esInicioSesion = false;  

        }

        private void Centrar()
        {

            this.nombrePrograma = this.Text;
            this.CenterToScreen();
            //this.Opacity = .97; //Está bien perro esto.  
            //this.Location = Screen.PrimaryScreen.WorkingArea.Location;
            //this.Size = Screen.PrimaryScreen.WorkingArea.Size;

        }

        private void AsignarTooltips()
        {

            ToolTip tp = new ToolTip();
            tp.AutoPopDelay = 5000;
            tp.InitialDelay = 0;
            tp.ReshowDelay = 100;
            tp.ShowAlways = true; 
            tp.SetToolTip(this.btnEntrar, "Entrar.");
            tp.SetToolTip(this.btnSalir, "Salir.");
            tp.SetToolTip(this.btnAyuda, "Ayuda.");

        }

        private void AsignarTooltips(string texto)
        {

            lblDescripcionTooltip.Text = texto;

        }

        private void AsignarFocos()
        {

            this.btnEntrar.Focus();            

        }

        private void ConfigurarConexionPrincipal() 
        {

            if (this.esDesarrollo)
            {
                baseDatos.CadenaConexionPrincipal = "C:\\Berry Agenda-Bitacora\\Principal.sdf";
            }
            else
            {
                string ruta = System.IO.Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().GetName().CodeBase);
                ruta = ruta.Replace("file:\\", null);
                baseDatos.CadenaConexionPrincipal = string.Format("{0}\\Principal.sdf", ruta); 
            }
            baseDatos.AbrirConexionPrincipal();
        
        }

        private void ConfigurarConexiones() 
        { 
             
            if (this.esDesarrollo)
            {
                baseDatos.CadenaConexionPrincipal = "C:\\Berry Agenda-Bitacora\\Principal.sdf";
                //Logica.DatosEmpresaPrincipal.instanciaSql = "ANDREW-MAC\\SQLEXPRESS";
                Logica.DatosEmpresaPrincipal.instanciaSql = "BERRY1-DELL\\SQLEXPRESS2008";
                Logica.DatosEmpresaPrincipal.usuarioSql = "AdminBerry";
                //Logica.DatosEmpresaPrincipal.contrasenaSql = "@berry";
                Logica.DatosEmpresaPrincipal.contrasenaSql = "@berry2017";
                Logica.DatosEmpresaPrincipal.idEmpresa = 1; 
                ConfigurarConexionPrincipal(); 
            }
            else
            { 
                ConfigurarConexionPrincipal();
                ConsultarInformacionEmpresaPrincipalPredeterminada(); 
            }   

        }
         
        private void ConsultarInformacionEmpresaPrincipalPredeterminada()
        {
             
            List<Entidades.EmpresasPrincipal> lista = new List<Entidades.EmpresasPrincipal>();
            lista = empresasPrincipal.ObtenerPredeterminada();
            Logica.DatosEmpresaPrincipal.idEmpresa = Convert.ToInt32(lista[0].IdEmpresa);
            Logica.DatosEmpresaPrincipal.activa = Convert.ToBoolean(lista[0].Activa.ToString());
            Logica.DatosEmpresaPrincipal.instanciaSql = Convert.ToString(lista[0].InstanciaSql.ToString());
            Logica.DatosEmpresaPrincipal.rutaBd = lista[0].RutaBd.ToString();
            Logica.DatosEmpresaPrincipal.usuarioSql = lista[0].UsuarioSql.ToString();
            Logica.DatosEmpresaPrincipal.contrasenaSql = lista[0].ContrasenaSql.ToString();

        }
                 
        private void CargarTitulosEmpresa()
        {

            this.Text = this.nombrePrograma;

        }

        private void CargarEncabezados()
        { 

            lblEncabezadoPrograma.Text = "Programa: " + this.nombrePrograma; 

        }

        private void Activar()
        {

            licencia.EsPrueba = false;
            licencia.FechaRegistro = DateTime.Today;
            licencia.FechaVencimiento = DateTime.Today;
            licencia.Eliminar();
            licencia.Guardar();

        }

        private void Desactivar()
        {

            licencia.EsPrueba = true;
            licencia.FechaRegistro = DateTime.Today;
            licencia.FechaVencimiento = DateTime.Today;
            licencia.Eliminar(); 

        }

        #endregion

    }
}